spec/models/odk/response_saver_spec.rb

```diff
@@ -55,28 +50,20 @@
         end
 
         thread2 = Thread.new do
-          # wait until the first thread is sleepin
-          sleep(3)
-          insert_response_via_second_db_connection(checksum)
+          ODK::ResponseSaver.save_with_retries!(
+            response: r2,
+            submission_file: upload,
+            user_id: user.id
+          )
         end
 
         thread1.join
         thread2.join
       rescue ActiveRecord::SerializationFailure => e
+        expect(Response.count).to eq(1)
         expect(ActiveStorage::Blob.all.count).to eq(1)
         expect(e.class).to eq(ActiveRecord::SerializationFailure)
       end
     end
   end
-
-  def insert_response_via_second_db_connection(checksum)
-    db = ActiveRecord::Base.connection_pool.checkout
-    db.execute("start transaction isolation level serializable;")
-    db.execute("select * from active_storage_blobs WHERE checksum='#{checksum}';")
-    db.execute("INSERT INTO active_storage_blobs
-      (checksum, byte_size, content_type, filename, key, service_name, created_at)
-      VALUES ('#{checksum}', 10000, 'application/xml', 'simple_response.xml', 'abc123', 'test', NOW());")
-    db.execute("COMMIT;")
-    ActiveRecord::Base.connection_pool.checkin(db)
-  end
```

Comment:
this is amazing and I want to save it for future usage, it will live on in the commit history -- the spec seems to pass as you wrote it even without this trick, but please advise if I'm missing a key ingredient.

Reply:
it passes for me too lol, i am NOT sure. well glad it is simplified then! I think in getting here, I wanted to ensure that i was creating the conditions for this type of collision. but it is just fine to know it works without and that we are getting the error we want!